### PR TITLE
Add run task results/stages and update for post-plan

### PR DESCRIPTION
**Pull request by glennsarti**
Previously the docs were missing the Task Stage and Task Result API
endpoints.  This commit adds the documentation and reworks the API
docs hierarchy for Run Tasks to be easier to navigate.

This commit also updates the Run API with the new post-plan states
and relationships for task-stages.

---

This commit retrospectively updates the API Changelog for parts of the Run
Tasks API endpoints.

This commit also fixes some whitespace and spelling issues.

### CODE REVIEW

**Comment by glennsarti** [comment]
@laurapacilio My branch with the additional Run Task docs

**Comment by laurapacilio** [comment]
Ah yes, I see exactly the issue. We need to make that more concise. What if we did something like:

Run Tasks
- CRUD Actions
- Stages
- Results
- Integration

I know we've never done this before, but to be fair, very few things are nested in the API docs. I feel like we need some way to be like "this page has your basic create and manage run tasks endpoints" and then the rest of the words make sense. My thought is that CRUD is very recognizable to developers and thus would be okay, but please push back on me here. I remember that Alexandra a while back was trying to do something similar for workspaces (that we will hopefully be able to implement someday) and she took a poll and developers felt favorably about "CRUD". 

If we can do something like this, then I don't think we need Run Tasks or even Tasks in front of the sidebar items, as they're under the Run Tasks heading in the sidebar. The page names themselves would still be "Run Tasks CRUD Actions" and "Run Task Stages API". This would help retain SEO and make sure the page contents are clear if users don't get to them using the sidebar. 

Let me know what you think. Open to feedback and discussion!

**Comment by glennsarti** [comment]
I will always defer to your team on the hierarchy decisions.  Personally, I am not a fan of using CRUD literally. We `list`, `show` and `associate` too.  I was looking at the Github API reference docs for inspiration. 

image

```
Run Tasks    v
    Run Tasks
    Task Stages
    Task Results
    Creating an Integration   <--- not sure about this name
```

**Comment by laurapacilio** [comment]
I value your opinion @glennsarti !  That's fair about CRUD...I knew it was a long shot, but I thought I'd propose it anyway. I like using GitHub as inspiration, and I see that they repeat the grouping header (organizations) so that makes me feel better about just having a page called "Run Tasks" under the "Run Tasks" grouping header.

In terms of the rest....I'm looking more at the actual content you've added, and I'm wondering if we could combine Stages and Results into the same page? They are both short, and it seems like they are related to one another. Because of this, maybe users would benefit from an overview that explains both and then provides the endpoints. I see that the overview you have on each page is very similar!

Then I was thinking about how you added more to integration and I get it --> we want to make sure people understand that this is where you go if you're trying to build your own run task for TFC. We don't use "ing" words in the API page names as of now, and though I think that could change in the future, I'm hesitant to add it as a one-off here. I'm wondering if we could say "custom integration". We use that term a lot on this page: https://www.terraform.io/cloud-docs/integrations/run-tasks#hcp-packer-run-task and I was thinking it may convey this idea that you know, this is for people who are building a new run task. 

What do you think of this below? All of the names are distinct and don't start with the same word (making them easier to skim). It's also hopefully clearer which one is related to straight up adding/removing/getting run tasks and which one is for folks who are trying to integrate.

Run Tasks
- Run Tasks
- Stages and Results
- Custom Integration

Once we settle on the high-level organization, I'll go ahead and do a very thorough review of the actual copy. Thank you for your work on this!

**Comment by glennsarti** [comment]
Ooo... I like that!

```
Run Tasks
    Run Tasks
    Stages and Results
    Custom Integration
```

I'll rework my PR with those changes and ping you for a review.  Thanks 🙏

**Comment by glennsarti** [comment]
Hierarchy change

image

File: website/docs/cloud-docs/api-docs/run-tasks/run-task-stages-and-results.mdx
Comment (cam-stitt) [inline website/docs/cloud-docs/api-docs/run-tasks/run-task-stages-and-results.mdx:62]: the trailing `s` outside of the JSON API document array seems odd.

File: website/docs/cloud-docs/api-docs/run-tasks/run-task-stages-and-results.mdx
Comment (cam-stitt) [inline website/docs/cloud-docs/api-docs/run-tasks/run-task-stages-and-results.mdx:257]: Nit: we use tables to display the states for Task Stages and Task Results, but a list for this. Should we use table here for consistency?

File: website/docs/cloud-docs/api-docs/run-tasks/run-task-stages-and-results.mdx
Comment (cam-stitt) [reply website/docs/cloud-docs/api-docs/run-tasks/run-task-stages-and-results.mdx:62]: Screen Shot 2022-06-14 at 1 18 18 pm

Comment (glennsarti) [reply website/docs/cloud-docs/api-docs/run-tasks/run-task-stages-and-results.mdx:62]: Huh so..

1. How on earth did you spot that? 🦅👀!
2. This is a copy-pasta from run and ssh-keys docs.  They have the same trailing 's'

Looks like it's needed to have a consistent reference to 'JSON API Document'.  But I'll change that to use 'documents' appropriately.

Comment (cam-stitt) [reply website/docs/cloud-docs/api-docs/run-tasks/run-task-stages-and-results.mdx:62]: I think `Array of [JSON API document][]` is fine. Just remove the s.

File: website/docs/cloud-docs/api-docs/run-tasks/run-task-stages-and-results.mdx
Comment (glennsarti) [reply website/docs/cloud-docs/api-docs/run-tasks/run-task-stages-and-results.mdx:257]: Hrmm... not saying it's right but this is format in other documents like run. Could probably be a table. I'll change it.

File: website/docs/cloud-docs/api-docs/run-tasks/run-task-stages-and-results.mdx
Comment (glennsarti) [reply website/docs/cloud-docs/api-docs/run-tasks/run-task-stages-and-results.mdx]: Changed to have a singular and plural link. Reads better with an s

File: website/docs/cloud-docs/api-docs/run.mdx
Comment (jarrodn-au) [inline website/docs/cloud-docs/api-docs/run.mdx:71]: Is this second 'in progress' accurate?

Comment (glennsarti) [reply website/docs/cloud-docs/api-docs/run.mdx:71]: no it is not

Comment (glennsarti) [reply website/docs/cloud-docs/api-docs/run.mdx]: Fixed

**Review by jarrodn-au — APPROVED** [review]
Good one 👍

**Review by laurapacilio — CHANGES_REQUESTED** [review]
Hi! Looking good! I like the new hierarchy. :-) 

I'm struggling a little bit and I think it's because I'm confused about the relationship between the various objects involved. Please ignore edits where I've misunderstood how things are connected. But! My overall thought is that we need to do more in the page introduction to explain Terraform run stages vs. Run Task stages vs. the states associated with Run Task stages, vs. Run Task results. I took a stab at this, but please feel free to edit as is needed.

Finally, I see now why you are using "Task Stages" etc. It's because the attributes are named that way. But, in titles, etc. on this page, I think we need to use the full feature name, which is "run task". I just want it to always be very clear what feature we're referring to, and "Task" is very generic. Let me know if you strongly disagree!
__________________

There are also a couple of to dos:
- We'll need to open a PR in `terraform-website` to add redirects for the pages we've renamed :-) Otherwise, users will get a 404 and be sad. You can do this in the `redirects.next.js` file. Once we do that, we can link that PR in this one's description so they're connected.
- Can we add language to the top of the Run Tasks page and the Integration page to link them to this new page/explain what's on it? Something like "Refer to the Run Task Stages and Results API for endpoints to retrieve run task states and details." My feeling is that all of the pages should provide pointers to each other in case users accidentally click one that doesn't have the exact endpoints they need.

**Comment by glennsarti** [comment]
@laurapacilio 

Ok let's see if I can explain this in long form:

Note - I don't address the Run Task Custom Integrations API calls. It's not relevant for the examples I give.

### Managing Run Tasks

These are the relationships when managing Run Tasks:

- A user creates a Run Task in the Organization (for the sake of being specific, this is an Organization Level Run Task)

- A user can then associate the Organization Level Run Task to zero or more Workspaces. These are known as Workspace Level Run Tasks.
  A Workspace level Run Tasks has an enforcement level (Advisory or Mandatory) and WHEN the task should be actioned within a Run (This is the stage property; pre-plan, post-plan etc.)

Unfortunately because they're both called "Tasks" or "Run Tasks", so it can cause some confusion which is why I refer to them as Org. or Workspace Run Tasks.


### So what happens when a Run occurs?

- A user creates a Run on a workspace (Either by the website, Terrafrom CLI or whatever...)

TFC creates the Run and all required related objects/things; An empty Plan, an empty Apply, Cost Estimation Results, Policy Check results, Task Stages and Task Results. This is BEFORE the run is even queued/actioned.

### So how are the stages and results related?

When the Run is created, it evaluates all of the Workspace Level Run Tasks; and creates a Task Stage for all of the unique stages (pre-plan, post-plan etc.)

It then creates a Task Result in each stage for every applicable Workspace Level Run Task.

Therefore:

- A Run can have zero or more Task Stages.
- A Task Stage can have one or more Task Results.

---

Here's a full example:

- Say we have 3 Organization Level Run Tasks; alpha, beta and gamma.
- We associate all 3 tasks to a Workspace;  alpha is assigned to the pre-plan stage, while beta and gamma are assigned to the post-plan stage.

- We create a Run
    - The Run has **two** Task Stages (pre-plan and post-plan)
    - The pre-plan Task Stage has **one** Task Result (for the alpha task)
    - The post-plan Task Stage has **two** Task Results (beta and gamma task)

File: website/docs/cloud-docs/api-docs/run-tasks/run-task-stages-and-results.mdx
Comment (glennsarti) [reply website/docs/cloud-docs/api-docs/run-tasks/run-task-stages-and-results.mdx:22]: I'll commit this as this.  For what it's worth, we will be adding a new stage very soon so that list will come back very quickly.

File: website/docs/cloud-docs/api-docs/run-tasks/run-task-stages-and-results.mdx
Comment (glennsarti) [reply website/docs/cloud-docs/api-docs/run-tasks/run-task-stages-and-results.mdx:37]: I'll rework that some more.

File: website/docs/cloud-docs/api-docs/run-tasks/run-task-stages-and-results.mdx
Comment (glennsarti) [reply website/docs/cloud-docs/api-docs/run-tasks/run-task-stages-and-results.mdx:45]: I really need a markdown linter to scream at me if don't terminate sentences with periods.

File: website/docs/cloud-docs/api-docs/run-tasks/run-task-stages-and-results.mdx
Comment (glennsarti) [reply website/docs/cloud-docs/api-docs/run-tasks/run-task-stages-and-results.mdx:47]: copy-pasta fail.  I'll change that.

File: website/docs/cloud-docs/api-docs/run-tasks/run-task-stages-and-results.mdx
Comment (glennsarti) [reply website/docs/cloud-docs/api-docs/run-tasks/run-task-stages-and-results.mdx:51]: unreachable happens when a stage/result could not be executed in a Run because the Run was stopped prior.

For example. A pre-apply Stage/Result is unreachable if the Run had no changes and was never applied;  That is the run was finished prior to executing the pre-apply phase, so the pre-apply never had the chance to execute.

File: website/docs/cloud-docs/api-docs/run.mdx
Comment (glennsarti) [reply website/docs/cloud-docs/api-docs/run.mdx:71]: fixed

File: website/docs/cloud-docs/api-docs/run-tasks/run-task-stages-and-results.mdx
Comment (glennsarti) [reply website/docs/cloud-docs/api-docs/run-tasks/run-task-stages-and-results.mdx:41]: I'm going to rewrite the whole top introduction which _should_ mean this description is no longer required.

**Comment by glennsarti** [comment]
Okay, I've added a commit which expands the task stage/result introduction.  I think given the back-n-forth and size of the text here, we may need to do zoom call to throw around ideas.

**Review by laurapacilio — APPROVED** [review]
I'm going to approve so I don't block, but I think it's definitely better! Thank you for taking the time to add the explanation to the top. I think it's helpful and important because these aren't concepts that actually correspond with anything in the UI. So I want to make sure folks understand what's getting created, why, and how.

Also thank you very much for typing that explanation to me! :)

File: website/docs/cloud-docs/api-docs/run-tasks/run-task-stages-and-results.mdx
Comment (glennsarti) [reply website/docs/cloud-docs/api-docs/run-tasks/run-task-stages-and-results.mdx]: Done

**Comment by glennsarti** [comment]
@laurapacilio   What's the order of operations here as I haven't raised the redirects in the terraform docs site yet.

Do we merge here and then TF docs? or TF docs and then merge here?

**Comment by laurapacilio** [comment]
Thank you for asking!  So what I would do is:
- open the redirects PR in `terraform-website`
- link that PR in this PR's description
- When approved, merge the `terraform-website` PR
- Merge this PR immediately after.

That way, we have a record in the PR description that we did indeed care about redirects AND when you merge this in and it becomes live, the redirects are in place and everything is (hopefully!) working perfectly :)

**Comment by glennsarti** [comment]
Redirects PR - https://github.com/hashicorp/terraform-website/pull/2355

**Pull request merged**
